scripts/nlp_language_modeling/build_knn_map_index.py

```diff
@@ -61,37 +100,48 @@
 
 
 @njit(parallel=True)
-def build_map(chunk_start, result, total_chunks):
+def build_map(chunk_start, result, total_chunks, start_id, end_id):
     """
-    build the map from chunk_id to document id
+    Build the map from chunk_id to a range of chunk ids that are from the same document.
+    The chunk_id is in range [start_id, end_id)
     """
     size = len(chunk_start)
     for i in prange(size):
         beg = chunk_start[i]
         end = chunk_start[i + 1] if i < size - 1 else total_chunks
-        result[beg:end] = i
+        if start_id < end and beg < end_id:  # [beg, end) intersect  [start_id, end_id)
+            result[max(beg - start_id, 0) : (end - start_id), 0] = beg
+            result[max(beg - start_id, 0) : (end - start_id), 1] = end
 
 
 @njit(parallel=True)
-def dedup(chunk_id_to_doc_id_map, I, tmp_neighbors, chunk_id_start):
+def _dedup(chunk_id_to_range, I, tmp_neighbors, chunk_id_start, offset):
+    for cid in prange(len(I)):
+        if chunk_id_start + cid - offset >= 0 and chunk_id_start + cid - offset < len(chunk_id_to_range):
+            beg, end = chunk_id_to_range[chunk_id_start + cid - offset]
+        position = 0
+        for target_chunk_id in I[cid]:
+            if beg <= target_chunk_id < end:
+                # target chunk is from the same document
+                continue
+            tmp_neighbors[cid, position] = target_chunk_id
+            position += 1
+
+
+def dedup(chunk_id_to_range, I, tmp_neighbors, chunk_id_start, offset):
     """
     deduplicate the KNN who are from the same document as the data chunks.
```

Comment:
Why do we want to do this? Can we not have multiple neighbors from a given document? Or am I misunderstanding something here?

Reply:
see above